### DISCUSSION TITLE
Limit CPU TDP for gaming on iGPU

### DISCUSSION THREAD

**Discussion by ethanp919** [seerge/g-helper#720] [Ideas]
I was hoping the CPU power limits would do this, but they don't seem to limit the overall TDP of the CPU in this way. This is useful for gaming on iGPU as you can set the TDP to something lower like 12 or 15w and play indie games for hours. There is another program that does this https://github.com/project-sbc/Handheld-Control-Panel/releases/tag/1.0.0.0 and it works perfectly, but it would be great to have this option as another slider in the CPU power limit section of G-Helper.

**Comment by seerge** [comment]
@ethanp919 hello. I have tried to answer you on reddit, but you seemed to ignore that.

_The way to limit CPU handheld panel is using is also already in the app, but it's being used only for old devices (from 2020 that don't support limits normally)._

Since you have 2023 model, it already supports CPU power limits in an "asus" way.  Slider in the app should work for you (as it works for other users with same devices). So I would prefer to figure out what happens in your specific case : 

1. Can you post log and config from ``%appdata/ghelper`` here ?
2. How do you verify if limit is working or not? Can you run cinebench23 and post a screenshot from HWInfo under load here (with some limits set)?
3. What's your BIOS version? Make sure to update to latest 310, it seem to fix a lot of things 
4. Does CPU limits work for you in Armoury Crate?

**Reply by ethanp919** [reply]
Hey sorry I didn't realize that was you on reddit. I just now updated BIOS to 310 and no difference. I am using HWINFO64 and RTSS in game to display CPU Package Power. Using G-Helper and applying an 11 watt limit shows the CPU Package Power in the 30w range. Doing the same steps in Handheld Panel the Package Power will remain at 11 watts. I did a clean install of Windows when I purchased this laptop so I do not have Armoury Crate installed. I do have all updates installed from the updates tab in G-Helper as well and Windows is up to date. 

[log.txt](https://github.com/seerge/g-helper/files/11950688/log.txt)
<img width="1280" alt="Screenshot 2023-07-04 111244" src="https://github.com/seerge/g-helper/assets/82677137/37b8dc6d-1d37-4a5e-96df-d7866db9a5f7">

**Reply by seerge** [reply]
@ethanp919 oke,

I have asked about Armoury cause wanted to check if your device would follow when limits are set there (as G-Helper uses Armoury's endpoints for limiting for all devices, including new Rog Ally and they work there too)

In log I see that they are accepted by BIOS
```
7/4/2023 11:10:24 AM: Mode = 2 : OK
7/4/2023 11:10:25 AM: PowerLimit A0 = 11 : OK
7/4/2023 11:10:25 AM: PowerLimit A3 = 11 : OK
7/4/2023 11:10:25 AM: PowerLimit C1 = 11 : OK
7/4/2023 11:10:25 AM: UV: -15 OK
7/4/2023 11:10:25 AM: iGPU UV: 0 CMD_REJECTED_PREREQ
7/4/2023 11:10:25 AM: CPU Temp: 96 OK
7/4/2023 11:10:25 AM: APU Temp: 96 OK
7/4/2023 11:10:36 AM: Mode = 2 : OK
```

In AC, "Manual" mode is actually Turbo with extra settings on top of it. Can you try to a following test :

1. Use turbo mode
2. Apply custom limits there + apply fan curve there
3. Don't set any custom undervolting / temp limits

Then run Cinebench23 (or any other CPU benchmark) and watch metrics in HWinfo. The value affected should be CPU PPT.
See my screenshot from my run on 2022 model for limit of 40W. Keep in mind that APU / CPU power package are actually higher cause they also include iGPU / etc. But CPU PPT strictly follows a limit.

![Screenshot 2023-07-04 185125](https://github.com/seerge/g-helper/assets/5920850/eea341d4-2266-42bc-be88-560b74c42f29)

**Reply by seerge** [reply]
P.S. What version of Asus Control interface do you have ? 
![Screenshot 2023-07-01 200836](https://github.com/seerge/g-helper/assets/5920850/7416c2d2-cfce-44ae-b089-b69b1bc2090e)

**Reply by ethanp919** [reply]
Okay so great news! Apparently I was not selecting the "Apply Custom Fan Curve" tab. I was selecting the power limits and applying those and selecting the fan curves from the top drop down menu, but I didn't have the bottom box clicked to apply them. Once I did that the power limits are working properly in all modes. I'm so glad to not need the other software now haha. Are you still working on GPU undervolting in upcoming patch? Once that is implemented this software will be all that is needed. Awesome work man!

**Reply by seerge** [reply]
@ethanp919 generally speaking, you didn't had to. Power limits can be set independently from fans. But on some models BIOS is bugged, and needs both (as in AC you can set them only together in "manual" mode).

What BIOS version do you have ? I think they have fixed a lot of stuff in latest 310

**Reply by ethanp919** [reply]
BIOS is updated to 310, but I didn't see any difference in this regard.

**Reply by seerge** [reply]
Can you check if power limits work w/o applying fans (in updated BIOS)

**Reply by seerge** [reply]
@Metalogic999 you also have 2023 model. Do power limits work for you when apply only them but not fans?

(I mainly want to figure out if it's an issue / limitation only with 4060 model or also for all 2023 devices ...)

**Reply by ethanp919** [reply]
Power limits are still not applying without custom fan curve checked on 310.  My model is 4070

**Reply by seerge** [reply]
@ethanp919 oke. 

This build will apply fans along power limits even if you don't check "apply" (there is a list of models that misbehave in the app, and I have added GA402XI there). It's just strange, as it's quite a lot of users with 2023 models, and no one complained about this so far ...

**[GHelper.zip](https://github.com/seerge/g-helper/files/11958755/GHelper.zip)**

**Reply by ethanp919** [reply]
Thanks works great!
